<center><img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="300"></center>


<h1>Data Wrangling with Tidyverse</h1>


Estimated Time Needed: **40 min**

<h3>Welcome!</h3>

By the end of this notebook, you will have learned the basics of Data Wrangling! Data Wrangling is the process of converting data from the initial format to a format that may be better for analysis.


### Table of Contents:
* [1. Missing Values and Formatting](#missing_values)
    * [1.1 Identify Missing Values](#identify_missing_values)
    * [1.2 Handle Missing Values](#handle_missing_values)
    * [1.3 Correct Data Format](#correct_data_format)
* [2. Data Normalization](#data_normalization)
    * [2.1 Simple Scaling](#simple_scale)
    * [2.2 Min-max](#min_max)
    * [2.3 Standardization (Z-score)](#data_standardization)
* [4. Binning](#binning)
* [5. Indicator Variable](#indicator)


## Load Libraries and Data


First, load the tidyverse library. 


In [1]:
# Load tidyverse
library(tidyverse)

── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──
✔ ggplot2 3.3.0     ✔ purrr   0.3.4
✔ tibble  3.0.1     ✔ dplyr   0.8.5
✔ tidyr   1.0.2     ✔ stringr 1.4.0
✔ readr   1.3.1     ✔ forcats 0.5.0
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()


The original Airline dataset is hosted on <a href="https://developer.ibm.com/exchanges/data/all/airline/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDA0151ENSkillsNetwork926-2022-01-01">IBM Data Asset eXchange</a>. This sample dataset can be found <a href="https://dax-cdn.cdn.appdomain.cloud/dax-airline/1.0.1/lax_to_jfk.tar.gz">here</a>.  We will be using a subset of the original dataset, which contains just LAX to JFK flights, throughout this course. 

Now using the subset dataset link, you can load it and store as a dataframe `sub_airline`:


In [2]:
# url where the data is located
url <- "https://dax-cdn.cdn.appdomain.cloud/dax-airline/1.0.1/lax_to_jfk.tar.gz"

# download the file
download.file(url, destfile = "lax_to_jfk.tar.gz")

# untar the file so we can get the csv only
# if you run this on your local machine, then can remove tar = "internal" 
untar("lax_to_jfk.tar.gz", tar = "internal")

# read_csv only 
sub_airline <- read_csv("lax_to_jfk/lax_to_jfk.csv",
                     col_types = cols('DivDistance' = col_number(), 
                                      'DivArrDelay' = col_number()))

Warning message in untar2(tarfile, files, list, exdir, restore_times):
“using pax extended headers”

<a id="missing_values"></a>
## 1. Missing Values and Formatting

Now that we have the data loaded, let's first take a look at the data using the method `head()`. 


In [3]:
head(sub_airline)

Month DayOfWeek FlightDate Reporting_Airline Origin Dest CRSDepTime
1  3    5         2003-03-28 UA                LAX    JFK  2210      
2 11    4         2018-11-29 AS                LAX    JFK  1045      
3  8    5         2015-08-28 UA                LAX    JFK  0805      
4  4    7         2003-04-20 DL                LAX    JFK  2205      
5 11    3         2005-11-30 UA                LAX    JFK  0840      
6  4    1         1992-04-06 UA                LAX    JFK  1450      
  CRSArrTime DepTime ArrTime ⋯ ArrDelayMinutes CarrierDelay WeatherDelay
1 0615       2209    0617    ⋯ 2               NA           NA          
2 1912       1049    1851    ⋯ 0               NA           NA          
3 1634       0757    1620    ⋯ 0               NA           NA          
4 0619       2212    0616    ⋯ 0               NA           NA          
5 1653       0836    1640    ⋯ 0               NA           NA          
6 2308       1452    2248    ⋯ 0               NA           NA          
  NASDelay SecurityDelay LateAircraftDelay DepDelay DepDelayMinutes DivDistance
1 NA       NA            NA                -1       0               NA         
2 NA       NA            NA                 4       4               NA         
3 NA       NA            NA                -8       0               NA         
4 NA       NA            NA                 7       7               NA         
5 NA       NA            NA                -4       0               NA         
6 NA       NA            NA                 2       2               NA         
  DivArrDelay
1 NA         
2 NA         
3 NA         
4 NA         
5 NA         
6 NA

As we can see, several NA (not available) appeared in the dataframe; those are missing values which may hinder our further analysis. 
So, how do we identify all those missing values and deal with them? 

In the following sections, we will go over the steps for working with missing data:
- Identify missing data
- Deal with missing data
- Correct data format


<a id="identify_missing_values"></a>
### 1.1 Identify Missing Malues

In R, there are some special symbols to represent special cases in data:
* `NA`: missing values are represented by the symbol `NA` (not available), it is a special symbol in R. Note, that `"NA"` (a string) is not the same as `NA`. 
* `NaN`: Impossible values (e.g., dividing by zero) are represented by the symbol `NaN` (not a number).

The missing values in airline dataset are already represented with R's `NA` symbol. We use R's built-in (also called **base R**) functions to identify these missing values. There are two methods to detect missing data:
<ol>
    <li><code>is.na(x)</code>: x can be a vector or list, this method returns a vector of TRUE or FALSE depending if the according element in x is <code>NA</code> or not. For example <code>is.na(c(1, NA))</code> returns <code>FALSE TRUE</code></li>
    <li><code>anyNA(x, recursive = FALSE)</code>: x can be a vector or list, this method returns TRUE if x contains any NAs and FALSE otherwise. For example <code>is.na(c(1, NA))</code> returns <code>TRUE</code></li>
</ol>


In [4]:
is.na(c(1, NA))        #> FALSE  TRUE
is.na(paste(c(1, NA))) #> FALSE FALSE

[1] FALSE  TRUE

[1] FALSE FALSE

Again, the output for `is.na()` is a vector of logical values where`TRUE` stands for missing value, while `FALSE` stands for not missing value.


In [5]:
anyNA(c(1, NA))   

[1] TRUE

The output for `anyNA()` is a vector of logical values where `TRUE` stands for at least one missing value in the vector, while `FALSE` stands for no missing values in the vector.


#### Counting Missing Values
    
We can quickly figure out the number of missing values in each column. As mentioned above, when using the function `is.na()`, `TRUE` represents a missing value while `FALSE` is otherwise.  The method  `sum()`  counts the number of `TRUE` values. 
    
Let's check how many missing values in `CarrierDelay` column and also check how many missing values in each column:


In [6]:
# counting missing values
sub_airline %>%
  summarize(count = sum(is.na(CarrierDelay)))

count
1 2486

We can use `purrr::map()` to count missing values in each of the columns. 

`map()` essentially maps (applies) a function or formula to each given element.
In the code below, it is mapping a *formula*, `~sum(is.na(.))`, that sums the NAs to every column in `sub_airline`. Since it is using a *formula*, you will also notice two special operators dot `.` and tilde `~`:
* The tilde `~` separates the left side of a formula with the right side. Normally formulas are two-sided like `y ~ x`, in this case in `map()`, the formula gets converted to a function so it only needs the right side.
* The dot `.` refers to each column in the dataset. If you view the documentation with `?map`, you can see that you use `.` when the function takes in just one parameter (a column in this case).

See `?formula` and `?map` for more information.


In [7]:
map(sub_airline, ~sum(is.na(.)))

$Month
[1] 0

$DayOfWeek
[1] 0

$FlightDate
[1] 0

$Reporting_Airline
[1] 0

$Origin
[1] 0

$Dest
[1] 0

$CRSDepTime
[1] 0

$CRSArrTime
[1] 0

$DepTime
[1] 0

$ArrTime
[1] 0

$ArrDelay
[1] 0

$ArrDelayMinutes
[1] 0

$CarrierDelay
[1] 2486

$WeatherDelay
[1] 2486

$NASDelay
[1] 2486

$SecurityDelay
[1] 2486

$LateAircraftDelay
[1] 2486

$DepDelay
[1] 0

$DepDelayMinutes
[1] 0

$DivDistance
[1] 2855

$DivArrDelay
[1] 2855

In [8]:
# Check dimensions of the dataset
dim(sub_airline)

[1] 2855   21

Based on the summary above, "CarrierDelay", "WeatherDelay", "NASDelay", "SecurityDelay" and "LateAircraftDelay" columns have 2486 rows of missing data, while "DivDistance" and "DivArrDelay" columns have 2855 rows of missing data. All other columns do not have missing data.
<ol>
    <li>"CarrierDelay": 2486 missing data</li>
    <li>"WeatherDelay": 2486 missing data</li>
    <li>"NASDelay": 2486 missing data</li>
    <li>"SecurityDelay" : 2486 missing data</li>
    <li>"LateAircraftDelay": 2486 missing data</li>
    <li>"DivDistance": 2855 missing data</li>
    <li>"DivArrDelay": 2855 missing data</li>
</ol>


<a id="handle_missing_values"></a>
### 1.2 Handle Missing Data
**How to deal with missing data?**

<ol>
    <li>Drop data<br>
        a. Drop the whole column<br>
        b. Drop the whole row
    </li>
    <li>Replace data<br>
        a. Replace it by mean<br>
        b. Replace it by frequency<br>
        c. Replace it based on other functions
    </li>
</ol>


Generally, you should not blindly drop NAs. However if an entire column or almost the entire column contains NAs, then it may be a good idea to leave it out. In our dataset, columns `DivDistance` and `DivArrDelay` are nearly all empty so we will drop them entirely.

**Drop the whole column:**
<ul>
    <li>"DivDistance": 2855 missing data</li>
    <li>"DivArrDelay": 2855 missing data</li>
</ul>


In [9]:
drop_na_cols <- sub_airline %>% select(-DivDistance, -DivArrDelay)
dim(drop_na_cols)
head(drop_na_cols)

[1] 2855   19

Month DayOfWeek FlightDate Reporting_Airline Origin Dest CRSDepTime
1  3    5         2003-03-28 UA                LAX    JFK  2210      
2 11    4         2018-11-29 AS                LAX    JFK  1045      
3  8    5         2015-08-28 UA                LAX    JFK  0805      
4  4    7         2003-04-20 DL                LAX    JFK  2205      
5 11    3         2005-11-30 UA                LAX    JFK  0840      
6  4    1         1992-04-06 UA                LAX    JFK  1450      
  CRSArrTime DepTime ArrTime ArrDelay ArrDelayMinutes CarrierDelay WeatherDelay
1 0615       2209    0617      2      2               NA           NA          
2 1912       1049    1851    -21      0               NA           NA          
3 1634       0757    1620    -14      0               NA           NA          
4 0619       2212    0616     -3      0               NA           NA          
5 1653       0836    1640    -13      0               NA           NA          
6 2308       1452    2248    -20      0               NA           NA          
  NASDelay SecurityDelay LateAircraftDelay DepDelay DepDelayMinutes
1 NA       NA            NA                -1       0              
2 NA       NA            NA                 4       4              
3 NA       NA            NA                -8       0              
4 NA       NA            NA                 7       7              
5 NA       NA            NA                -4       0              
6 NA       NA            NA                 2       2

**Drop the whole row:**
<ul>
    <li>"CarrierDelay": 2486 missing data</li>
    <li>"WeatherDelay": 2486 missing data</li>
    <li>"NASDelay": 2486 missing data</li>
    <li>"SecurityDelay" : 2486 missing data</li>
    <li>"LateAircraftDelay": 2486 missing data</li>   
</ul>

We see `CarrierDelay`, `WeatherDelay`, `NASDelay`, `SecurityDelay`, `LateAircraftDelay` have the same amount of missing values from the summary. By dropping the missing values in one column will also solve the missing value issues in the others.


In [10]:
# Drop the missing values
drop_na_rows <- drop_na_cols %>% drop_na(CarrierDelay)
dim(drop_na_rows)
head(drop_na_rows)

[1] 369  19

Month DayOfWeek FlightDate Reporting_Airline Origin Dest CRSDepTime
1  8    5         2018-08-03 B6                LAX    JFK  0013      
2  6    4         2006-06-01 AA                LAX    JFK  1515      
3  1    7         2007-01-28 UA                LAX    JFK  0845      
4  6    5         2013-06-28 AA                LAX    JFK  1200      
5  9    1         2010-09-27 DL                LAX    JFK  1330      
6 10    3         2005-10-12 AA                LAX    JFK  0930      
  CRSArrTime DepTime ArrTime ArrDelay ArrDelayMinutes CarrierDelay WeatherDelay
1 0839       0034    0913    34       34              11           0           
2 2332       1507    2353    21       21               0           0           
3 1656       0838    1713    17       17               0           0           
4 2045       1328    2220    95       95               5           0           
5 2208       1426    2316    68       68               0           0           
6 1755       0958    1823    28       28               0           0           
  NASDelay SecurityDelay LateAircraftDelay DepDelay DepDelayMinutes
1 23       0              0                21       21             
2 21       0              0                -8        0             
3 17       0              0                -7        0             
4  7       0             83                88       88             
5 68       0              0                56       56             
6 28       0              0                28       28

We have some freedom in choosing which method to replace data; however, some methods may seem more reasonable than others. In this scenario, we would like to replace missing values with 0.

**Convert NA to 0**

In the airline dataset, missing data for the different types of delay corresponds to no delay. So, we can replace these NAs with 0 in this case.
To do this we use the function: 
 <pre>tidyr::replace_na(data, replace, ...)</pre>

The columns that corresponds with types of delays are CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, and LateAircraftDelay. For example
if `CarrierDelay = NA` then this means there is no delay in Carrier, so the delay in minutes can be changed to 0 or `CarrierDelay = 0`. Let's transform these columns and see the result:


In [11]:
# Replace the missing values in five columns
replace_na <- drop_na_rows %>% replace_na(list(CarrierDelay = 0,
                                              WeatherDelay = 0,
                                              NASDelay = 0,
                                              SecurityDelay = 0,
                                              LateAircraftDelay = 0))
head(replace_na)

Month DayOfWeek FlightDate Reporting_Airline Origin Dest CRSDepTime
1  8    5         2018-08-03 B6                LAX    JFK  0013      
2  6    4         2006-06-01 AA                LAX    JFK  1515      
3  1    7         2007-01-28 UA                LAX    JFK  0845      
4  6    5         2013-06-28 AA                LAX    JFK  1200      
5  9    1         2010-09-27 DL                LAX    JFK  1330      
6 10    3         2005-10-12 AA                LAX    JFK  0930      
  CRSArrTime DepTime ArrTime ArrDelay ArrDelayMinutes CarrierDelay WeatherDelay
1 0839       0034    0913    34       34              11           0           
2 2332       1507    2353    21       21               0           0           
3 1656       0838    1713    17       17               0           0           
4 2045       1328    2220    95       95               5           0           
5 2208       1426    2316    68       68               0           0           
6 1755       0958    1823    28       28               0           0           
  NASDelay SecurityDelay LateAircraftDelay DepDelay DepDelayMinutes
1 23       0              0                21       21             
2 21       0              0                -8        0             
3 17       0              0                -7        0             
4  7       0             83                88       88             
5 68       0              0                56       56             
6 28       0              0                28       28

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  #1: </h1>

<b>According to the example above, let's try to replace NA in "CarrierDelay" column by the mean value.</b>
</div>


In [12]:
# Write your code below and press Shift+Enter to execute 
# Calculate the mean vaule for "CarrierDelay" column
carrier_mean <- mean(drop_na_rows$CarrierDelay)

# Replace NA by mean value in "CarrierDelay" column
sub_airline %>% replace_na(list(CarrierDelay = carrier_mean))

Month DayOfWeek FlightDate Reporting_Airline Origin Dest CRSDepTime
1     3    5         2003-03-28 UA                LAX    JFK  2210      
2    11    4         2018-11-29 AS                LAX    JFK  1045      
3     8    5         2015-08-28 UA                LAX    JFK  0805      
4     4    7         2003-04-20 DL                LAX    JFK  2205      
5    11    3         2005-11-30 UA                LAX    JFK  0840      
6     4    1         1992-04-06 UA                LAX    JFK  1450      
7    12    1         2012-12-03 VX                LAX    JFK  1040      
8    12    3         2004-12-01 HP                LAX    JFK  2200      
9     2    4         1997-02-13 UA                LAX    JFK  0800      
10    3    4         2016-03-10 DL                LAX    JFK  2245      
11   10    4         1999-10-14 AA                LAX    JFK  2300      
12    8    5         2018-08-03 B6                LAX    JFK  0013      
13    6    4         2006-06-01 AA                LAX    JFK  1515      
14    2    3         1999-02-03 AA                LAX    JFK  0700      
15    8    7         1995-08-27 AA                LAX    JFK  0900      
16    3    6         1990-03-31 PA (1)            LAX    JFK  0830      
17    6    4         1996-06-13 AA                LAX    JFK  0700      
18    7    4         2003-07-24 AA                LAX    JFK  2315      
19   12    5         2018-12-07 AS                LAX    JFK  1530      
20    1    4         1996-01-25 DL                LAX    JFK  2300      
21    7    2         1992-07-21 AA                LAX    JFK  2300      
22    8    3         1990-08-29 TW                LAX    JFK  1600      
23    4    1         2016-04-11 DL                LAX    JFK  1255      
24    1    7         2007-01-28 UA                LAX    JFK  0845      
25    1    2         2000-01-18 UA                LAX    JFK  0800      
26    6    5         2013-06-28 AA                LAX    JFK  1200      
27    9    4         2015-09-03 AA                LAX    JFK  1515      
28    5    4         1997-05-15 UA                LAX    JFK  1300      
29    8    2         2010-08-24 DL                LAX    JFK  0800      
30   11    4         1999-11-25 TW                LAX    JFK  1230      
⋮    ⋮     ⋮         ⋮          ⋮                 ⋮      ⋮    ⋮         
2826  9    1         1997-09-22 DL                LAX    JFK  0940      
2827  8    2         2001-08-28 UA                LAX    JFK  1625      
2828  4    1         2016-04-04 DL                LAX    JFK  0610      
2829  5    5         1995-05-26 UA                LAX    JFK  0825      
2830  3    4         2010-03-18 DL                LAX    JFK  1240      
2831  2    6         1995-02-11 AA                LAX    JFK  0800      
2832  3    6         2004-03-27 DL                LAX    JFK  2200      
2833  3    7         2020-03-01 AA                LAX    JFK  1700      
2834  2    3         2016-02-17 AA                LAX    JFK  1330      
2835 12    2         1998-12-01 UA                LAX    JFK  1430      
2836  6    5         2018-06-01 DL                LAX    JFK  0615      
2837  2    1         2015-02-16 AA                LAX    JFK  2330      
2838  9    3         2018-09-05 B6                LAX    JFK  1342      
2839  1    1         1996-01-29 UA                LAX    JFK  1115      
2840  1    3         2016-01-06 VX                LAX    JFK  1345      
2841 10    4         2002-10-03 AA                LAX    JFK  2200      
2842  3    2         2016-03-15 VX                LAX    JFK  0735      
2843 11    7         1998-11-29 UA                LAX    JFK  1600      
2844  1    7         1990-01-14 AA                LAX    JFK  1200      
2845  1    1         2007-01-01 UA                LAX    JFK  1600      
2846  5    6         1988-05-14 AA                LAX    JFK  0830      
2847  1    7         1994-01-02 UA                LAX    JFK  0730      
2848  4    1         2000-04-17 UA                LAX    JFK  0

<details>
    <summary>Click here for the solution.</summary>

```r
# Calculate the mean vaule for "CarrierDelay" column
carrier_mean <- mean(drop_na_rows$CarrierDelay)

# Replace NA by mean value in "CarrierDelay" column
sub_airline %>% replace_na(list(CarrierDelay = carrier_mean))
```
</details>


<b>Good!</b> Now, we obtain the dataset with no missing values.


<a id="correct_data_format"></a>
### 1.3 Correct Data Format
**We are almost there!**

<p>The last step in data cleaning is checking and making sure that all data is in the correct format.</p>

#### Scoped dplyr Verbs
In dplyr, you can add `_all` and `_if` to the end of it's main functions like `mutate`, `filter`, `group_by`, and `summarize`. 

| dply function  | _all           | _if             |
| ------------- |:--------------:| ----------------:|
| mutate         | `mutate_all()`  |`mutate_if()`     |
| filter         | `filter_all()`  |`filter_if()`     |
| group_by       |`group_by_all()` |`group_by_if()`   |
| summarize      | `summarize_all()`  | `summarize_if()` |

These are called **scoped** dplyr verbs. The `_all` variant applies an operation on *all* variaibles and the `_if` variant applies an operation to a variable if the given function is `TRUE`.


Now that we understand these different versions of dplyr functions, let's list the data types for each column by using `summarize_all()` and `gather()`:


In [13]:
sub_airline %>% 
    summarize_all(class) %>% 
    gather(variable, class)

variable          class    
1  Month             numeric  
2  DayOfWeek         numeric  
3  FlightDate        Date     
4  Reporting_Airline character
5  Origin            character
6  Dest              character
7  CRSDepTime        character
8  CRSArrTime        character
9  DepTime           character
10 ArrTime           character
11 ArrDelay          numeric  
12 ArrDelayMinutes   numeric  
13 CarrierDelay      numeric  
14 WeatherDelay      numeric  
15 NASDelay          numeric  
16 SecurityDelay     numeric  
17 LateAircraftDelay numeric  
18 DepDelay          numeric  
19 DepDelayMinutes   numeric  
20 DivDistance       numeric  
21 DivArrDelay       numeric

<p>It is important for later analysis to explore the feature’s data type and convert them to the correct data types; otherwise, the developed models later on may behave strangely, and totally valid data may end up being treated like missing data.</p> 

Let's re-format the `FlightDate` field into three separate fields (year, month, day) using `separate()`.


In [ ]:
date_airline <- replace_na %>% 
    separate(FlightDate, sep = "-", into = c("year", "month", "day"))

head(date_airline)

For a number of reasons, including when you import a dataset into R or process a variable, the data type may be incorrectly established. For example, here we notice that the assigned data type to for `year`, `month`, and `day` is “character” although the expected data type should really be numeric. You can change the type using `mutate_all()` and `mutate_if()`. 

In the below code, it is mutating `year`, `month`, and `day` values to numeric only if it is a character. Note that in `mutate_if()`, the first parameter is the function `is.character` that checks a *condition* while the second parameters is the function `as.numeric` that *modifies* the data *if* the condition is true.


In [ ]:
date_airline %>%
    select(year, month, day) %>%
    mutate_all(type.convert) %>%
    mutate_if(is.character, as.numeric)

<b>Wonderful!</b>

Now, we finally obtain the cleaned dataset with no missing values and all data in its proper format.


<a id="data_normalization"></a>

## 2. Data Normalization

**Why normalization?**

Normalization is the process of transforming values of several features (variables) into a similar range. An example of why this could be important is if you have a variable for income and another variable for age. Income is likely much bigger values than age, so in a model, income would naturally influence the model more. Thus, normalization helps make comparisons between different variables more *fair*.

There are different types of ways to normalize:
* **Simple scaling**: divides each value by the maximum value in a feature. The new range is between 0 and 1.
$$x_{new} = \frac{x_{old}}{x_{max}}$$
* **Min-max**: subtracts the minimum value from the original and divides by the maximum minus the minimum. The minimum becomes 0 and the maximum becomes 1.
$$x_{new} = \frac{x_{old} - x_{min}}{x_{max} - x_{min}}$$
* **Standardization (Z-score)**: subtract the mean ($\mu$) of the feature and divide by the standard deviation ($\sigma$).
$$x_{new} = \frac{x_{old} - \mu}{\sigma}$$

From these methods, you can see that typically normalizations include scaling the variable so that the range is from 0 to 1.


<a id="simple_scale"></a>
### 2.1 Simple scaling

**Simple scaling** divides each value by the maximum value in a feature. The new range is between 0 and 1.
$$x_{new} = \frac{x_{old}}{x_{max}}$$

**Example**

To demonstrate simpling scaling, let's say we want to scale the column "ArrDelay".
    
* **Target:** Would like to Normalize those variables so their value ranges from 0 to 1.
* **Approach:** Replace the original value by (original value)/(maximum value)


In [ ]:
simple_scale <- sub_airline$ArrDelay / max(sub_airline$ArrDelay)
head(simple_scale)

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #2: </h1>

<b>According to the example above, normalize the column "DepDelay" using the simple scaling technique.</b>
</div>


In [ ]:
# Write your code below and press Shift+Enter to execute 


<details>
    <summary>Click here for the solution.</summary>

```r
simple_scale2 <- sub_airline$DepDelay/max(sub_airline$DepDelay)
head(simple_scale2)
```
                 
</details>


<a id="min_max"></a>
### 2.2 Min-max

**Min-max** subtracts the minimum value from the original and divides by the maximum minus the minimum. The minimum becomes 0 and the maximum becomes 1.
$$x_{new} = \frac{x_{old} - x_{min}}{x_{max} - x_{min}}$$

**Example**

Using "ArrDelay" as an example again, you can transform this column using the min-max technique:


In [ ]:
minmax_scale <- (sub_airline$ArrDelay - min(sub_airline$ArrDelay)) /
                (max(sub_airline$ArrDelay) - min(sub_airline$ArrDelay))
head(minmax_scale)

<a id="data_standardization"></a>
### 2.3 Data Standardization (Z-score)

**Standardization (Z-score)** subtracts the mean ($\mu$) of the feature and divides by the standard deviation ($\sigma$).
$$x_{new} = \frac{x_{old} - \mu}{\sigma}$$


**Example**

Let's use "ArrDelay" again. We can use `mean()` to find the mean of the feature and we can use `sd()` to find the standard deviation of the feature.


In [ ]:
z_scale <- (sub_airline$ArrDelay - mean(sub_airline$ArrDelay)) / sd(sub_airline$ArrDelay)
head(z_scale)

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  #3: </h1>

<b>According to the example above, standardize the "DepDelay" column.</b>
</div>


In [ ]:
# Write your code below and press Shift+Enter to execute 


<details>
    <summary>Click here for the solution.</summary>

```r
z_scale2 <- (sub_airline$DepDelay - mean(sub_airline$DepDelay)) / sd(sub_airline$DepDelay)
head(z_scale2)
```
                 
</details>


<a id="binning"></a>
## 4. Binning
### Why use binning?
<p>
    Binning is a process of transforming continuous numerical variables into discrete categorical 'bins', for grouped analysis.
</p>

**Example:**
<p>Using binning, we categorize arrival delays into four bins by quartiles. Quartiles refer to the boundaries that divide observations into four defined intervals. They are often determined based on the values of data points and how they are compared with the rest of the dataset. In the actual airline dataset, "ArrDelay" is a numerical variable ranging from -73 to 682, it has 2855 unique values. We can categorize them into 4 bins. Can we rearrange them into four ‘bins' to simplify analysis? </p>

We will use the tidyverse method `mutate(quantile_rank = ntile())` to segment the "ArrDelay" column into 4 bins


### Example of Binning Data In Tidyverse


Lets plot the histogram of flight arrival delays, to see what the distribution of "ArrDelay" looks like.


In [ ]:
ggplot(data = sub_airline, mapping = aes(x = ArrDelay)) +
  geom_histogram(bins = 100, color = "white", fill = "red") +
  coord_cartesian(xlim = c(-73, 682))

First we use the dplyr function `ntile` to break “ArrDelay” into 4 buckets, which have equal amount of observations of flight arrival delays. We then create a list “quantile_rank” that contains 4 bins, which are respectively labeled "1", "2", "3", "4". So bin 1 would contain the first 25% of data, bin 2 the next 25% of data and so on.


In [ ]:
binning <- sub_airline %>%
      mutate(quantile_rank = ntile(sub_airline$ArrDelay,4))

head(binning)

The observations are put into different bins based on the flights' delay minutes. The larger the bin label is, the longer the flight was delayed. 

Now if we look at a histogram of the bins, you can see that all bins are equal.


In [ ]:
ggplot(data = binning, mapping = aes(x = quantile_rank)) +
  geom_histogram(bins = 4, color = "white", fill = "red")

<a id="indicator"></a>
## 5. Indicator variable
#### What is an indicator variable?

An indicator variable (or dummy variable) is a **numerical variable** used to **label categories**. They are called 'dummies' because the numbers themselves don't have inherent meaning. 


**Why we use indicator variables?**
<p>
Regression models need numerical variables, however categorical variables in their original forms are strings. Using indicator variables allows us to be able to use these categorical variables in regression models.
</p>

**Example**
<p>
    In the airline dataset, the "Reporting_Airline" feature is a categorical variable that has nine values, "AA", "AS", "B6", "DL", "HP", "PA (1)", "TW", "UA" or "VX”, which are in character type. For further analysis, we need to convert these variables into some form of numeric format. 
</p>


We will use the tidyverse's method `spread()` method to convert categorical variables to dummy variables. The parameters to use in the function are:
* `key`: the column to convert into categorical values
* `value`: the value you want to set the key to
* `fill`: fills the missing values with this value

Also keep in mind that the method will drop the "Reporting_Airline" column by default. So if you used "ArrDelay" as the key value instead, that column would get dropped.

The below code does the following:
1. `mutate` - creates a column `dummy` with all 1's then
2. `spread` - creates new column for every `Reporting_Ariline` and sets the value to 1 from `dummy`. But replaces the 1 with a 0 if the corresponding value is `NA` in `Reporting_Airline` then
3. `slice` - looks at the specified rows


In [ ]:
sub_airline %>%
  mutate(dummy = 1) %>% # column with single value
  spread(
    key = Reporting_Airline, # column to spread
    value = dummy,
    fill = 0) %>%
  slice(1:5)

When a value occurs in the original feature, we set the corresponding value to one in the new feature; the rest of the features are set to zero. 

So in the output above, for the first row, the reporting airline is "UA". Therefore, the feature "UA" is set to one and the other features to zero. 
Similarly, for the second row, the reporting airline value is "AS". Therefore, the feature "AS" is set to one and the other features to zero.


Alternatively, instead of assigning dummy values 0 or 1, we can assign flight delay values to each feature. So this will also create new columns, one for each `Reporting_Airline`. Taking the first row for example, the column "UA" is now set to 2 because that is the value from ArrDelay.


In [ ]:
sub_airline %>% 
  spread(Reporting_Airline, ArrDelay) %>% 
  slice(1:5)


**Visualize Airline Category**

Let's visualize how many data points in each airline category.


In [ ]:
sub_airline %>% # start with data
   mutate(Reporting_Airline = factor(Reporting_Airline,
                                     labels = c("AA", "AS", "DL", "UA", "B6", "PA (1)", "HP", "TW", "VX")))%>%
  ggplot(aes(Reporting_Airline)) +
  stat_count(width = 0.5) +
  labs(x = "Number of data points in each airline")

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  #4: </h1>

<b>As above, create indicator variable to the column of "Month".</b>
</div>


In [ ]:
# Write your code below and press Shift+Enter to execute


<details>
    <summary>Click here for the solution.</summary>

```r
sub_airline %>%
  mutate(dummy = 1) %>% # column with single value
  spread(
    key = Month, # column to spread
    value = dummy,
    fill = 0) %>%
  slice(1:5)
```
                 
</details>


 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  #5: </h1>

<b>Now, create indicator variable to the column of "Month" by applying departure delay values</b>
</div>


In [ ]:
# Write your code below and press Shift+Enter to execute 


<details>
    <summary>Click here for the solution.</summary>

```r
sub_airline %>% 
  spread(Month, DepDelay) %>% 
  slice(1:5) # Show only the first 5 rows
```
                 
</details>


<h1>Thank you for completing this notebook</h1>


<h3>About the Authors:</h3>

This notebook was written by <a href="https://www.linkedin.com/in/yiwen-li-47a019119/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDA0151ENSkillsNetwork926-2022-01-01" target="_blank">Yiwen Li</a> and <a href="https://www.linkedin.com/in/gabrieladequeiroz/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDA0151ENSkillsNetwork926-2022-01-01" target="_blank">Gabriela de Queiroz</a>.

<p><a href="https://www.linkedin.com/in/yiwen-li-47a019119/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDA0151ENSkillsNetwork926-2022-01-01" target="_blank">Yiwen Li</a> has approximately three year experiences in big tech industry. Currently, she is a developer advocate, a data scientist, a product manager at IBM, where she designs and develops data science solutions and Machine Learning models to solve real world problems. She has delivered talks this year in JupyterCon, PyCon, Pyjamas, CrowdCast.ai, Global AI on Tour 2020 and Belpy 2021 with hundreds of attendants per talk. 
    
<a href="https://www.linkedin.com/in/gabrieladequeiroz/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDA0151ENSkillsNetwork926-2022-01-01" target="_blank">Gabriela de Queiroz</a> is a Sr. Engineering & Data Science Manager at IBM where she manages and leads a team of developers working on Data & AI Open Source projects. She works to democratize AI by building tools and launching new open source projects.
She is the founder of AI Inclusive, a global organization that is helping increase the representation and participation of gender minorities in Artificial Intelligence. She is also the founder of R-Ladies, a worldwide organization for promoting diversity in the R community with more than 190 chapters in 50+ countries.
She has worked in several startups and where she built teams, developed statistical models, and employed a variety of techniques to derive insights and drive data-centric decisions


Copyright © 2021 IBM Corporation. All rights reserved.
